File: packages/js-dpp/lib/dataContract/stateTransition/DataContractUpdateTransition/validation/state/validateDataContractUpdateTransitionStateFactory.js
const ValidationResult = require('../../../../../validation/ValidationResult');

const InvalidDataContractVersionError = require('../../../../../errors/consensus/basic/dataContract/InvalidDataContractVersionError');
const DataContractNotPresentError = require('../../../../../errors/consensus/basic/document/DataContractNotPresentError');
const Identity = require('../../../../../identity/Identity');
const IdentityPublicKey = require('../../../../../identity/IdentityPublicKey');
const InvalidSignaturePublicKeyIdError = require('../../../../../errors/consensus/state/identity/InvalidSignaturePublicKeyIdError');

/**
 *
 * @param {StateRepository} stateRepository
 * @return {validateDataContractUpdateTransitionState}
 */
function validateDataContractUpdateTransitionStateFactory(
  stateRepository,
) {
  /**
   * @typedef validateDataContractUpdateTransitionState
   * @param {DataContractCreateTransition} stateTransition
   * @return {ValidationResult}
   */
  async function validateDataContractUpdateTransitionState(stateTransition) {
    const result = new ValidationResult();

    const dataContract = stateTransition.getDataContract();
    const dataContractId = dataContract.getId();

    // Data contract should exist
    const existingDataContract = await stateRepository.fetchDataContract(dataContractId);

    if (!existingDataContract) {
      result.addError(
        new DataContractNotPresentError(dataContractId.toBuffer()),
      );

      return result;
    }

    // Version difference should be exactly 1
    const oldVersion = existingDataContract.getVersion();
    const newVersion = dataContract.getVersion();
    const versionDiff = newVersion - oldVersion;

    if (versionDiff !== 1) {
      result.addError(
        new InvalidDataContractVersionError(
          oldVersion + 1,
          oldVersion + versionDiff,
        ),
      );
    }

    if (!result.isValid()) {
      return result;
    }

    const identityId = stateTransition.getIdentityId();
    const storedIdentity = await stateRepository.fetchIdentity(identityId);

    // copy identity
    const identity = new Identity(storedIdentity.toObject());

    if (stateTransition.getBIP16Script()) {
      const publicKey = identity.getPublicKeyById(stateTransition.getSignaturePublicKeyId());
      if (publicKey.getType() !== IdentityPublicKey.TYPES.BIP13_SCRIPT_HASH) {
        result.addError(
          new InvalidSignaturePublicKeyIdError(stateTransition.getSignaturePublicKeyId()),
        );

        return result;
      }
    }

    return result;
  }

  return validateDataContractUpdateTransitionState;
}

module.exports = validateDataContractUpdateTransitionStateFactory;

File: packages/rs-dpp/src/data_contract/state_transition/data_contract_update_transition/validation/state/validate_data_contract_update_transition_state_factory.rs
use anyhow::Result;

use crate::{
    data_contract::{state_transition::DataContractUpdateTransition, DataContract},
    errors::consensus::basic::BasicError,
    state_repository::StateRepositoryLike,
    validation::ValidationResult,
};

pub struct DataContractUpdateTransitionStateValidator<SR>
where
    SR: StateRepositoryLike,
{
    state_repository: SR,
}

impl<SR> DataContractUpdateTransitionStateValidator<SR>
where
    SR: StateRepositoryLike,
{
    pub fn new(state_repository: SR) -> Self {
        DataContractUpdateTransitionStateValidator { state_repository }
    }

    pub async fn validate_data_contract_update_transition_state(
        &self,
        state_transition: &DataContractUpdateTransition,
    ) -> ValidationResult<()> {
        let mut result = ValidationResult::default();

        // Data contract should exist
        let maybe_existing_data_contract: Result<Option<DataContract>> = self
            .state_repository
            .fetch_data_contract(&state_transition.data_contract.id)
            .await;

        let existing_data_contract: DataContract = match maybe_existing_data_contract {
            // assumed the conservativeness for the validation. TBD: in the case of
            // general error we want to add the same result
            Ok(None) | Err(_) => {
                let err = BasicError::DataContractContPresent {
                    data_contract_id: state_transition.data_contract.id.clone(),
                };
                result.add_error(err);
                return result;
            }
            Ok(Some(dt)) => dt,
        };

        // Version difference should be exactly 1
        let old_version = existing_data_contract.version;
        let new_version = state_transition.data_contract.version;
        let version_diff = new_version - old_version;

        if version_diff != 1 {
            let err = BasicError::InvalidDataContractVersionError {
                expected_version: old_version + 1,
                version: old_version + version_diff,
            };
            result.add_error(err);
        }

        result
    }
}
